server/core/src/main/java/org/infinispan/server/core/backup/resources/CacheResource.java

```diff
@@ -0,0 +1,282 @@
+            ImmutableSerializationContext serCtx = ctxRegistry.getPersistenceCtx();
+            try (InputStream is = zip.getInputStream(zipEntry)) {
+               while (is.available() > 0) {
+                  CacheBackupEntry entry = readMessageStream(serCtx, CacheBackupEntry.class, is);
+                  Object key = keyMarshalling ? unmarshall(entry.key, userMarshaller) : scm.getKeyWrapper().wrap(entry.key);
+                  Object value = valueMarshalling ? unmarshall(entry.value, userMarshaller) : scm.getKeyWrapper().wrap(entry.value);
+                  Metadata metadata = unmarshall(entry.metadata, persistenceMarshaller);
+                  Metadata internalMetadataImpl = new InternalMetadataImpl(metadata, entry.created, entry.lastUsed);
+
+                  PutKeyValueCommand cmd = commandsFactory.buildPutKeyValueCommand(key, value, keyPartitioner.getSegment(key),
+                        internalMetadataImpl, FlagBitSets.IGNORE_RETURN_VALUES);
+                  cmd.setInternalMetadata(entry.internalMetadata);
+                  invocationHelper.invoke(cmd, 1);
+               }
+            } catch (IOException e) {
+               throw new CacheException(e);
+            }
+         }, "restore-cache-" + cacheName));
+      }
+      return stages.freeze();
+   }
+
+   private CompletionStage<Void> createCacheBackup(String cacheName) {
+      return blockingManager.runBlocking(() -> {
+         AdvancedCache<?, ?> cache = cm.getCache(cacheName).getAdvancedCache();
+         Configuration configuration = cm.getCacheConfiguration(cacheName);
+
+         // Create the cache backup dir and parents
+         Path cacheRoot = root.resolve(cacheName);
+         mkdirs(cacheRoot);
+
+         // Write configuration file
+         String xmlFileName = configFile(cacheName);
+         Path xmlPath = cacheRoot.resolve(xmlFileName);
+         try (OutputStream os = Files.newOutputStream(xmlPath)) {
+            parserRegistry.serialize(os, cacheName, configuration);
+         } catch (XMLStreamException | IOException e) {
+            throw new CacheException(String.format("Unable to create backup file '%s'", xmlFileName), e);
+         }
+
+         // Write in-memory cache contents to .dat file if the cache is not empty
+         if (cache.isEmpty())
+            return;
+
+         ComponentRegistry cr = cache.getComponentRegistry();
+         ClusterPublisherManager<?, ?> clusterPublisherManager = cr.getClusterPublisherManager().running();
+         SerializationContextRegistry ctxRegistry = cr.getGlobalComponentRegistry().getComponent(SerializationContextRegistry.class);
+         ImmutableSerializationContext serCtx = ctxRegistry.getPersistenceCtx();
+
+         String dataFileName = dataFile(cacheName);
+         Path datFile = cacheRoot.resolve(dataFileName);
+
+         Publisher<CacheEntry<?, ?>> p = s -> clusterPublisherManager.entryPublisher(null, null, null, false,
```

Comment:
This is something that I have been in two minds about. Originally I was thinking that if data is already persisted, then the stores can also be migrated by the user, but this isn't very user-friendly especially if stores are not shared and we have a large cluster. Having a centralised backup greatly simplifies this, so including stores by default makes more sense on second thought.

In the case of a large shared store, probably JDBC based, it's probably not desired to have the store content in the backup. Once the core backup/restore pieces are in place (CLI, Operator etc) we can add a parameter to toggle this behaviour.

Reply:
Makes sense. This makes me wonder if we should have the default value be if the store it uses is shared or not.

Also I wonder if we should add support for the Flag `SKIP_SHARED_CACHE_STORE` in the `ClusterPublisherManager`

Reply:
> This makes me wonder if we should have the default value be if the store it uses is shared or not

Tbh I think this could be difficult to explain to the user in a transparent way. Having a consistent default and then documenting that it's possible to backup in-memory contents only is much simpler IMO. We can then add a note explaining that for larger shared stores it's probably desirable to enable this option.

Reply:
> Also I wonder if we should add support for the Flag `SKIP_SHARED_CACHE_STORE` in the `ClusterPublisherManager`

Makes sense. It would be perfect for this use-case